lib/code_corps/accounts/accounts.ex

```diff
@@ -29,8 +34,92 @@ defmodule CodeCorps.Accounts do
   def create_from_github_changeset(struct, %{} = params) do
     struct
     |> Changeset.change(params |> Adapters.User.from_github_user())
-    |> Changeset.put_change(:context, "github")
+    |> Changeset.put_change(:sign_up_context, "github")
     |> Changeset.unique_constraint(:email)
     |> Changeset.validate_inclusion(:type, ["bot", "user"])
   end
+
+  @doc ~S"""
+  Updates a user record using attributes from a GitHub payload along with the
+  access token.
+  """
+  @spec update_from_github_oauth(User.t, map, String.t) :: {:ok, User.t} | {:error, Changeset.t}
+  def update_from_github_oauth(%User{} = user, %{} = params, access_token) do
+    params =
+      params
+      |> Adapters.User.from_github_user()
+      |> Map.put(:github_auth_token, access_token)
+
+    changeset = user |> update_from_github_oauth_changeset(params)
+
+    multi =
+      Multi.new
+      |> Multi.update(:user, changeset)
+      |> Multi.run(:installations, fn %{user: %User{} = user} -> user |> associate_installations() end)
+      |> Multi.run(:tasks, fn %{user: %User{} = user} -> user |> associate_tasks() end)
+      |> Multi.run(:comments, fn %{user: %User{} = user} -> user |> associate_comments() end)
+
+    case Repo.transaction(multi) do
+      {:ok, %{user: %User{} = user, installations: installations}} ->
+        {:ok, user |> Map.put(:github_app_installations, installations)}
+      {:error, :user, %Changeset{} = changeset, _actions_done} ->
+        {:error, changeset}
+    end
+  end
+
+  @doc ~S"""
+  Casts a changeset used for creating a user account from a github user payload
+  """
+  @spec update_from_github_oauth_changeset(struct, map) :: Changeset.t
+  def update_from_github_oauth_changeset(struct, %{} = params) do
+    struct
+    |> Changeset.cast(params, [:github_auth_token, :github_avatar_url, :github_id, :github_username, :type])
+    |> Changeset.validate_required([:github_auth_token, :github_avatar_url, :github_id, :github_username, :type])
```

Comment:
Are we not setting the email at all then? 

My understanding was, we set the user email from github if there's no email already.

I can see the reasoning behind not wanting to set it at all, but I want to make sure.

Reply:
In this case, where the user is connecting their GitHub account via OAuth, the assumption is that there already is an email. I can see how maybe we would want to not assume that and pull it and set it if it's not yet set. I'm not quite sure how to do that, though.

Reply:
```Elixir
def ensure_email(%Changeset{} = changeset, %{"email" => email} = _params do
  case changeset |> Changeset.get_field(:email) do
    nil -> changeset |> Changeset.put_change(:email, email)
    _email -> changeset
  end
end
```

Something like that. `get_field/2` returns the value of the `get_change/2` if there is a change, or of the underlying struct field, if there is no change to it.

Basically, we set the email from github, only if the current email is nil and is not being changed to something else already. This covers both the case of using this changeset directly, as well as potentially using it as part of another changeset.

Reply:
@begedin I'm not sure I see how this works if someone has their email set in the database though. This only pulls from the changeset and params, no?

Reply:
It pulls from the changes, then from the database field. 

```Elixir
iex> %Changeset{data: %User{email: "foo"}, changes: %{email: "bar"}} |> get_field(:email)

"bar"

iex> %Changeset{data: %User{email: "foo"}, changes: %{}} |> get_field(:email)

"foo"